### PR TITLE
[T1006-MG < T1019-MG] Add LBA checks to all Create and Delete procedures in C API

### PR DESCRIPTION
**Pull request by BorisTasevski**

### CODE REVIEW

File: src/query/procedure/mg_procedure_impl.cpp [hunk 1]
```diff
@@ -1695,7 +1701,14 @@ mgp_error mgp_vertex_copy(mgp_vertex *v, mgp_memory *memory, mgp_vertex **result
   return WrapExceptions([v, memory] { return NewRawMgpObject<mgp_vertex>(memory, *v); }, result);
 }
 
-void mgp_vertex_destroy(mgp_vertex *v) { DeleteRawMgpObject(v); }
+void mgp_vertex_destroy(mgp_vertex *v) {
+  if (v->graph->ctx && v->graph->ctx->auth_checker &&
+      !v->graph->ctx->auth_checker->Accept(*v->graph->ctx->db_accessor, v->impl, v->graph->view /*, CREATE_DELETE*/)) {
+    return;
```
Comment (Josipmrden) [inline src/query/procedure/mg_procedure_impl.cpp:1707]: Do we throw an exception?

Comment (BorisTasevski) [reply src/query/procedure/mg_procedure_impl.cpp:1708]: No, we will expand return mgp_error enum and return that value

File: src/query/procedure/mg_procedure_impl.cpp [hunk 2]
```diff
@@ -1659,6 +1674,12 @@ mgp_error mgp_vertex_add_label(struct mgp_vertex *v, mgp_label label) {
 
 mgp_error mgp_vertex_remove_label(struct mgp_vertex *v, mgp_label label) {
   return WrapExceptions([=] {
+    if (v->graph->ctx && v->graph->ctx->auth_checker &&
```
Comment (Josipmrden) [inline src/query/procedure/mg_procedure_impl.cpp:1677]: ask in channel :)

File: tests/e2e/lba_procedures/create_delete_query_modules.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+from common import (
+    connect,
+    execute_and_fetch_all,
+    mgclient,
+    reset_create_delete_permissions,
+)
+
+create_vertex_query = "CALL create_delete.create_vertex() YIELD created_node RETURN labels(created_node);"
+remove_label_vertex_query = "CALL create_delete.remove_label('create_delete_label') YIELD node RETURN labels(node);"
+set_label_vertex_query = 'CALL create_delete.set_label("new_create_delete_label") YIELD node RETURN labels(node);'
```
Comment (Josipmrden) [inline tests/e2e/lba_procedures/create_delete_query_modules.py:23]: make consensus whether to use double or single quotes in strings

File: tests/e2e/lba_procedures/create_delete_query_modules.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+# included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+# License, and you may not use this file except in compliance with the Business Source License.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0, included in the file
+# licenses/APL.txt.
+
+import sys
+import pytest
```
Comment (Josipmrden) [inline tests/e2e/lba_procedures/create_delete_query_modules.py:13]: switch

Comment (Josipmrden) [reply tests/e2e/lba_procedures/create_delete_query_modules.py:13]: + newline before common

File: tests/e2e/lba_procedures/create_delete_query_modules.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+def test_can_not_create_vertex_when_given_global_update():
+    admin_cursor = connect(username="admin", password="test").cursor()
+    reset_create_delete_permissions(admin_cursor)
+
+    execute_and_fetch_all(admin_cursor, "GRANT UPDATE ON LABELS :create_delete_label TO user;")
+
+    test_cursor = connect(username="user", password="test").cursor()
+
+    with pytest.raises(mgclient.DatabaseError, match="AuthorizationError"):
```
Comment (Josipmrden) [inline tests/e2e/lba_procedures/create_delete_query_modules.py:71]: extract AuthorizationError into constant

File: tests/e2e/lba_procedures/update_query_modules.py [hunk 6]
```diff
@@ -9,8 +9,8 @@
 # by the Apache License, Version 2.0, included in the file
 # licenses/APL.txt.
 
-import sys
 import pytest
+import sys
 from common import (
```
Comment (Josipmrden) [inline tests/e2e/lba_procedures/update_query_modules.py:14]: newline

File: tests/e2e/lba_procedures/workloads.yaml [hunk 7]
```diff
@@ -61,8 +61,23 @@ show_privileges_cluster: &show_privileges_cluster
...
+      args: ["--bolt-port", "7687", "--log-level=TRACE"]
+      log_file: "lba-e2e.log"
+      setup_queries:
+        [
+          "CREATE USER admin IDENTIFIED BY 'test';",
+          "GRANT ALL PRIVILEGES TO admin",
+          "CREATE USER user IDENTIFIED BY 'test';",
+          "GRANT ALL PRIVILEGES TO user",
+        ]
+      validation_queries: []
```
Comment (Josipmrden) [inline tests/e2e/lba_procedures/workloads.yaml:78]: separate list items in yml with bullet dashes 
```
- "CREATE USER admin IDENTIFIED BY 'test';"
- "GRANT ALL PRIVILEGES TO admin"
- "CREATE USER user IDENTIFIED BY 'test';"
- "GRANT ALL PRIVILEGES TO user"
```

**Review by Josipmrden — CHANGES_REQUESTED** [review]
Some comments + please post in engineering channel the doubt with remove label

**Review by jbajic — APPROVED** [review]
Just one small comment, otherwise looks fine!

File: src/query/procedure/mg_procedure_impl.cpp [hunk 8]
```diff
@@ -1627,6 +1634,14 @@ mgp_error mgp_vertex_set_property(struct mgp_vertex *v, const char *property_nam
 
 mgp_error mgp_vertex_add_label(struct mgp_vertex *v, mgp_label label) {
   return WrapExceptions([=] {
+    if (v->graph->ctx && v->graph->ctx->auth_checker &&
+        !(v->graph->ctx->auth_checker->Accept(*v->graph->ctx->db_accessor, v->impl, v->graph->view,
+                                              memgraph::query::AuthQuery::FineGrainedPrivilege::UPDATE) &&
+          v->graph->ctx->auth_checker->Accept(*v->graph->ctx->db_accessor, {v->graph->impl->NameToLabel(label.name)},
+                                              memgraph::query::AuthQuery::FineGrainedPrivilege::CREATE_DELETE))) {
```
[the remark this answers is not in the document]
Comment (BorisTasevski) [reply src/query/procedure/mg_procedure_impl.cpp:1641]: Will be corrected in TechDebt task

File: include/mgp.py [hunk 9]
```diff
@@ -134,6 +134,15 @@ class SerializationError(_mgp.SerializationError):
     pass
 
 
+class AuthorizationError(_mgp.AuthorizationError):
+    """
+    Signals that the user doesn't have sufficient permissions to perform
+    procedure call.
+    """
+
+    pass
```
[the remark this answers is not in the document]
Comment (katarinasupe) [reply include/mgp.py:143]: Thank you, @jbajic, for tagging us!

**Pull request merged**